src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.LegacyEditor/RazorWrapperFactory.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.VisualStudio.Editor.Razor;
+
+namespace Microsoft.AspNetCore.Razor.ExternalAccess.LegacyEditor;
+
+/// <summary>
+///  Creates and caches wrappers for various Razor objects without directly exposing Razor types.
+/// </summary>
+[EditorBrowsable(EditorBrowsableState.Never)]
+internal static partial class RazorWrapperFactory
+{
+    private static readonly ConditionalWeakTable<object, object> s_objectToWrapperMap = new();
+
+    [return: NotNullIfNotNull(nameof(obj))]
+    private static TResult? Wrap<TInner, TWrapper, TResult>(object? obj, Func<TInner, TWrapper> createWrapper)
```

Comment:
Is there a reason the first param has to be `object?` rather than `TInner?`.

Reply:
This is `object?` to allow the helpers below to pass their arguments directly without have to introduce a type case.